### ISSUE TITLE
Different outputs for Onnx and .pt file

### ISSUE THREAD

**Issue by timcov** [ultralytics/ultralytics#22737]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hey, i have a good performing .pt file. But once i export it to onnx the predictions are completely different. 

train code
```
"""
Very simple training script for YOLO11 Segmentation models using the Ultralytics YOLO library.

Usage:
    python train.py
"""
from ultralytics import YOLO

# Load a model
model = YOLO("./modeling/yolo11s-seg.pt")  # load a pretrained model (recommended for training)

# Train the model
results = model.train(data="./configs/data.yaml ", epochs=100, imgsz=1024)
```

Export onnx:
```
from ultralytics import YOLO
from pathlib import Path

MODEL_PATH = "/runs/segment/train5/weights/best.pt"

model = YOLO(MODEL_PATH)

# Common, portable settings:
onnx_path = model.export(
    format="onnx",
    imgsz=(1024, 864),  
    # dynamic=True,     
)

print("Exported ONNX:", onnx_path)
print("Saved at:", Path(onnx_path).resolve())
```

Test code 
```
from ultralytics import YOLO

pt_model   = YOLO("/runs/segment/train5/weights/best.pt")
onnx_model = YOLO("/runs/segment/train5/weights/best.onnx") 

img = "/data/processed/yolo/data/validation/images/veld_9_2025_08_6_13_37_08_982.png"

print("Running inference on PT model..."    )
pt_res   = pt_model(img )[0]
print("Running inference on ONNX model..."  )
onnx_res = onnx_model(img)[0]    


    
print("Onnx vs PT results comparison:")
print(" PT boxes:", pt_res.boxes.xyxy)
print("Onnx boxes:", onnx_res.boxes.xyxy)
print(" PT masks:", pt_res.masks.data.shape if pt_res.masks is not None else None)
print("Onnx masks:", onnx_res.masks.data.shape if onnx_res.masks is not None else None)
```

Output 
```
Running inference on PT model...

image 1/1 /data/processed/yolo/data/validation/images/veld_9_2025_08_6_13_37_08_982.png: 1024x864 4 flaxs, 66.7ms
Speed: 7.7ms preprocess, 66.7ms inference, 5.9ms postprocess per image at shape (1, 3, 1024, 864)
Running inference on ONNX model...
Loading /runs/segment/train5/weights/best.onnx for ONNX Runtime inference...
WARNING ⚠️ Failed to start ONNX Runtime with CUDA. Using CPU...
Using ONNX Runtime 1.24.0.dev20251031003 CPUExecutionProvider

image 1/1 /data/processed/yolo/data/validation/images/veld_9_2025_08_6_13_37_08_982.png: 1024x864 2 flaxs, 160.1ms
Speed: 5.5ms preprocess, 160.1ms inference, 21.2ms postprocess per image at shape (1, 3, 1024, 864)
Onnx vs PT results comparison:
 PT boxes: tensor([[4.7519e+02, 1.0389e-01, 2.0269e+03, 8.5890e+02],
        [1.1509e+02, 0.0000e+00, 2.0115e+03, 1.6393e+03],
        [0.0000e+00, 3.8686e+01, 1.9636e+03, 2.3330e+03],
        [6.7010e+02, 8.8944e-01, 2.0215e+03, 5.4408e+02]], device='cuda:0')
Onnx boxes: tensor([[4.7514e+02, 1.0783e-01, 2.0269e+03, 8.5889e+02],
        [6.7007e+02, 8.9214e-01, 2.0215e+03, 5.4406e+02]])
 PT masks: torch.Size([4, 1024, 864])
Onnx masks: torch.Size([2, 1024, 864])
```

Less boxes

**Comment by UltralyticsAssistant** [comment]
👋 Hello @timcov, thanks for opening this issue and for sharing your scripts and logs 🚀  
This is an automated response to help gather initial details — an Ultralytics engineer will also assist you here soon 🙂

From your description, it looks like you are observing different outputs between a YOLO11 segmentation `.pt` model and its exported ONNX version (fewer boxes and masks in ONNX). To help us investigate efficiently:

- If this is a 🐛 **Bug Report**, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) that:
  - Uses a small, shareable image or dataset sample
  - Includes the exact `ultralytics` version, `Python` version, and `PyTorch` version
  - Includes the exact export and inference commands (or a minimal script) that reproduces the discrepancy
- If this is a custom training ❓ **Question**, please add:
  - A brief description of your dataset and task
  - Any relevant training logs
  - Confirmation that you have reviewed and followed the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

In the meantime, you may also want to review the general usage docs for reference:
- [Ultralytics Docs](https://docs.ultralytics.com/)
- [Python Usage Examples](https://docs.ultralytics.com/usage/python/)
- [CLI Usage Examples](https://docs.ultralytics.com/usage/cli/)

You’re very welcome to connect with the community for more real-time discussion and to compare behavior with others:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse](https://community.ultralytics.com/)
- [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Before we dive deeper, please ensure you have tested with the latest `ultralytics` release, as related issues may already be resolved there. Upgrade in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed:

```bash
pip install -U ultralytics
```

Retest your `.pt` vs ONNX comparison after upgrading and let us know if the behavior persists 💡

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

If possible, please also indicate:
- Which of the above environments you are using (or if it is a custom local setup)
- Whether the `.pt` vs ONNX difference appears in multiple environments or just one

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once you share the additional details and (ideally) a minimal reproducible example, it will greatly help the engineering team analyze the ONNX vs `.pt` behavior you are seeing 🙌

**Comment by Y-T-G** [comment]
Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by timcov** [comment]
Ultralytics 8.3.225 🚀 Python-3.12.3 torch-2.9.0+cu126 CUDA:0 (NVIDIA GeForce RTX 4070 Ti SUPER, 15941MiB)
Setup complete ✅ (28 CPUs, 125.5 GB RAM, 403.8/456.3 GB disk)

OS                     Linux-6.14.0-29-generic-x86_64-with-glibc2.39
Environment            Linux
Python                 3.12.3
Install                git
Path                   /media/covicon-thinktank/SSD_1TB/git/C25-0016-ZwadDetectieUnion/.venv/lib/python3.12/site-packages/ultralytics
RAM                    125.48 GB
Disk                   403.8/456.3 GB
CPU                    Intel Core i9-9940X 3.30GHz
CPU count              28
GPU                    NVIDIA GeForce RTX 4070 Ti SUPER, 15941MiB
GPU count              1
CUDA                   12.6

numpy                  ✅ 2.3.4>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 12.0.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.9.0+cu126>=1.8.0
torch                  ✅ 2.9.0+cu126!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.0+cu126>=0.9.0
psutil                 ✅ 7.1.3
polars                 ✅ 1.35.1
ultralytics-thop       ✅ 2.0.18>=2.0.18

**Comment by Y-T-G** [comment]
Can you show the output images?

**Comment by timcov** [comment]
@Y-T-G Here are the output images:
PT:

<img width="2048" height="2448" alt="Image" src="https://github.com/user-attachments/assets/a177c354-e3b9-43af-b9ec-a79168c70655" />

Onnx:

<img width="2048" height="2448" alt="Image" src="https://github.com/user-attachments/assets/b308973d-2616-4929-b715-407ffd4de045" />

**Comment by Y-T-G** [comment]
They were very likely removed due to the overlap. You can use `iou=0.8` during inference

**Comment by timcov** [comment]
@Y-T-G 

I changed the test code to this
```
print("Running inference on PT model..."    )
pt_res   = pt_model(img, conf=0.25, iou=0.8, imgsz=(1024, 864))[0]
print("Running inference on ONNX model..."  )
onnx_res = onnx_model(img, conf=0.25, iou=0.8, imgsz=(1024, 864))[0]
```
But gives the exact same results.

**Comment by Y-T-G** [comment]
What about 0.9?

**Comment by timcov** [comment]
No, gives me more boxes on the pt result but not anymore boxes on the onnx.

```
print("Running inference on PT model..."    )
pt_res   = pt_model(img, conf=0.25, iou=0.9, imgsz=(1024, 864))[0]
print("Running inference on ONNX model..."  )
onnx_res = onnx_model(img, conf=0.25, iou=0.9, imgsz=(1024, 864))[0]
```

PT:

<img width="2048" height="2448" alt="Image" src="https://github.com/user-attachments/assets/fe2e98f8-c32a-4572-9e9a-7d7d33226768" />

ONNX:

<img width="2048" height="2448" alt="Image" src="https://github.com/user-attachments/assets/b038a2b9-ff1b-439e-9306-5b4393bcdbc5" />

**Comment by Y-T-G** [comment]
Did you use `nms=True` during ONNX export?

**Comment by timcov** [comment]
Yes, it has nms=True. I now exported with False and it seems to work now. 

This is now conf 0.7 for pt and .9 for onnx

PT:

<img width="2048" height="2448" alt="Image" src="https://github.com/user-attachments/assets/d7bf6b3f-b7f7-49a4-8e5e-e845811b5f2a" />

ONNX:

<img width="2048" height="2448" alt="Image" src="https://github.com/user-attachments/assets/feff8cf7-16c7-465b-a5b3-a876a8cada95" />

**Comment by glenn-jocher** [comment]
Good find — that behavior is expected: with `nms=True` the ONNX graph has NMS baked in using the `conf`/`iou` values from export time, so changing `conf`/`iou` at inference can’t fully mirror the PyTorch post‑processing, whereas exporting with `nms=False` and letting Ultralytics handle NMS in Python (as you’re doing now) is the recommended approach when you want PT/ONNX parity; if you ever need NMS inside the graph, re‑export with the thresholds you want and see the [ONNX export docs](https://docs.ultralytics.com/integrations/onnx/) for more details.
